### PR TITLE
Object: Let debug lock handle callee destruction within call chain gracefully

### LINKED ISSUE #75658: Dereferencing a funcref from the function it calls leads to a heap-use-after-free
### Godot version

3.5.1.stable

### System information

Arch Linux (rolling release)

### Issue description

Dereferencing a funcref from the function it calls can cause an error in AddressSanitizer in debug/release-debug builds. It might be possible for this to cause more subtle heap corruption issues in release builds--I suspect it is behind a rare/unreproducible crash in my project.

I think what happens in the below example is that when `foobar` dereferences the funcref, the funcref is freed. Then when control flow returns from `foobar` to `FuncRef.call_func` the `this` pointer is invalid. It seems weird that function calls do not implicitly hold a reference to the object they're on.

```
==16930==ERROR: AddressSanitizer: heap-use-after-free on address 0x612000254d70 at pc 0x0000066b3b05 bp 0x7fff849419c0 sp 0x7fff849419b0
WRITE of size 4 at 0x612000254d70 thread T0
    #0 0x66b3b04 in std::__atomic_base<unsigned int>::fetch_sub(unsigned int, std::memory_order) /usr/include/c++/12.2.1/bits/atomic_base.h:628
    #1 0x66b3b04 in SafeNumeric<unsigned int>::decrement() core/safe_refcount.h:84
    #2 0x66b3b04 in SafeRefCount::unref() core/safe_refcount.h:177
    #3 0x66b3b04 in _ObjectDebugLock::~_ObjectDebugLock() core/object.cpp:53
    #4 0x66b3b04 in Object::call(StringName const&, Variant const**, int, Variant::CallError&) core/object.cpp:924
    #5 0x68b8c7a in Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) core/variant_call.cpp:1229
    #6 0x12f7960 in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1048
    #7 0x122592e in GDScriptInstance::call_multilevel(StringName const&, Variant const**, int) modules/gdscript/gdscript.cpp:1214
    #8 0x29b9039 in Node::_notification(int) scene/main/node.cpp:58
    #9 0x2cb073f in Node::_notificationv(int, bool) scene/main/node.h:47
    #10 0x2cb073f in CanvasItem::_notificationv(int, bool) scene/2d/canvas_item.h:163
    #11 0x2cb073f in Control::_notificationv(int, bool) scene/gui/control.h:47
    #12 0x66ab532 in Object::notification(int, bool) core/object.cpp:927
    #13 0x2a4c801 in SceneTree::_notify_group_pause(StringName const&, int) scene/main/scene_tree.cpp:1105
    #14 0x2a733d1 in SceneTree::idle(float) scene/main/scene_tree.cpp:608
    #15 0xb08189 in Main::iteration() main/main.cpp:2298
    #16 0xaaf46f in OS_X11::run() platform/x11/os_x11.cpp:3987
    #17 0xa4e5dc in main platform/x11/godot_x11.cpp:59
    #18 0x7f1c6d03c78f  (/usr/lib/libc.so.6+0x2378f)
    #19 0x7f1c6d03c849 in __libc_start_main (/usr/lib/libc.so.6+0x23849)
    #20 0xa64634 in _start (/media/scratch/coding/godot/godot-builder/godot/bin/godot.x11.opt.debug.64s+0xa64634)

0x612000254d70 is located 48 bytes inside of 272-byte region [0x612000254d40,0x612000254e50)
freed by thread T0 here:
    #0 0x7f1c6dcbe672 in __interceptor_free /usr/src/debug/gcc/gcc/libsanitizer/asan/asan_malloc_linux.cpp:52
    #1 0x6753795 in void memdelete<Reference>(Reference*) core/os/memory.h:118
    #2 0x6753795 in void memdelete<Reference>(Reference*) core/os/memory.h:110
    #3 0x6753795 in Ref<Reference>::unref() core/reference.h:258
    #4 0x6753795 in RefPtr::unref() core/ref_ptr.cpp:85

previously allocated by thread T0 here:
    #0 0x7f1c6dcbfa89 in __interceptor_malloc /usr/src/debug/gcc/gcc/libsanitizer/asan/asan_malloc_linux.cpp:69
    #1 0x6b16b2f in Memory::alloc_static(unsigned long, bool) core/os/memory.cpp:75
    #2 0x12f66ed in GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) modules/gdscript/gdscript_function.cpp:1116
    #3 0x123621b in GDScriptInstance::_ml_call_reversed(GDScript*, StringName const&, Variant const**, int) modules/gdscript/gdscript.cpp:1229
```

### Steps to reproduce

Put this script on a node and then play the scene, using a release_debug build with asan on:

```GDScript
var fr

func _ready():
	fr = funcref(self, "foobar")

func foobar():
	fr = null

func _process(_delta):
	if fr:
		fr.call_func()

```

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by RandomShaper**
Fixes #75658.

### CODE REVIEW

**Comment by AThousandShips** [comment]
How does this relate to:
* https://github.com/godotengine/godot/issues/86788

And this fix:
* https://github.com/godotengine/godot/pull/86969

**Comment by RandomShaper** [comment]
> How does this relate to:
> 
> * [Player crashes randomly in Godot 4.2 on macOS #86788](https://github.com/godotengine/godot/issues/86788)
> 
> And this fix:
> 
> * [[GDScript] Fix lifetime of call base object #86969](https://github.com/godotengine/godot/pull/86969)

Wow, I had totally forgotten I was involved there.

Now I understand the issue better, I can say that GDScript already ensures the lifetime of a `RefCounted` base object, because it stores it on the stack, so it will survive at least up to the end of the call. Therefore, I'd say there's no need for the extra temporaries added by the compiler in #86969.

**Comment by RandomShaper** [comment]
Of course, using `Variant` has the side effect of deferring the destruction of the object until after the call, so I've modified the approach in a way that the refcount is not artificially incremented, while still having safety if the object is released prior to returning.

**Comment by lawnjelly** [comment]
This looks pretty good and identical to my approach in the issue (great minds think alike and all that :grin: ), so am almost there to approve.

The only thing I wasn't sure about is how often this gets called in practice with every call to an object, and how expensive the ObjectDB lookup is (hashtable?) and whether we should be looking to avoid it, even in debug.

An alternative would be to optionally store e.g. a pointer to a bool on each `Object`, which can inform the calling code when an object has been destroyed. This is O(1) but has the disadvantage of storing an extra pointer on each object.

If the `Object::call()` isn't used that often it isn't worth micro-optimizing though.

**Review by lawnjelly — APPROVED** [review]

**Review by AThousandShips — APPROVED** [review]
Makes sense and looks good to me

**Comment by AThousandShips** [comment]
Would be good to add some tests if we can reliably reproduce these issues

**Comment by RandomShaper** [comment]
Regarding performance, I wouldn't worry much since `ObjectDB::get_instance()` is very fast, and O(1), as it's based on generational ids. On the other hand, I think there's an opportunity to revamp this debug lock feature and maybe make it work only if the debugger is attached (like certain `Variant` object validity checks do) and/or make a very different implementation. After all, the only goal of the debug lock is prevent `free()` from being called if the object is at that point involved in a call chain.

I've force-pushed with an added `likely()`, and also (due to the striking coincidence) including @lawnjelly as a co-author (that's the least a gentleman can do 😀).

I'm adding some tests and will push again.

**Comment by lawnjelly** [comment]
> I've force-pushed with an added likely(), and also (due to the striking coincidence) including @lawnjelly as a co-author (that's the least a gentleman can do 😀).

Ah yes I'll do that on mine too. :+1:

**Comment by RandomShaper** [comment]
Pushed some tests, as "black-boxey" as possible. There are other cases of `OBJ_DEBUG_LOCK` used in `object.cpp` (`emit_signalp` for the emitting object and `set_script()`), but in those it wouldn't be safe that the object destroyed itself.

**Comment by lawnjelly** [comment]
I don't know the first thing about tests, but given the MRP only crashes with asan and debug build, does the test fail with the old code (with the normal tests build)?

**Comment by RandomShaper** [comment]
> I don't know the first thing about tests, but given the MRP only crashes with asan and debug build, does the test fail with the old code (with the normal tests build)?

It will only fail with asan enabled, but luckily CI runs jobs that compile with sanitizers and runs tests on such builds.

**Comment by RandomShaper** [comment]
I've been able to come up with a way of letting a crash happen despite asan not enabled. Not as guaranteed, but makes the test crash with the prior implementation of `_ObjectDebugLock`.

**Comment by RandomShaper** [comment]
Wow, getting the non-asan/tsan version of the crash induction to work on every platform has proven challenging.

Also, I've added tests for GDScript as its script instance and VM implementation have to guarantee the behavior. Those will only fail with asan/tsan, though.

I'm fairly happy with this now.

File: core/object/object.cpp [hunk 1]
```diff
@@ -53,7 +53,7 @@ struct _ObjectDebugLock {
 	}
 	~_ObjectDebugLock() {
 		Object *obj_ptr = ObjectDB::get_instance(obj_id);
-		if (likely(obj_ptr)) {
+		if (obj_ptr) {
```
Comment (akien-mga) [inline core/object/object.cpp:56]: Did you intend to remove this? If so should it be part of the first commit?

Comment (RandomShaper) [reply core/object/object.cpp]: No, I didn't. Good catch!

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
Cherry-picked for 4.3.1.

**Pull request merged**